Stats Traffic date picker

guarani opened the pull request:
Fixes #22382

Add the Stats Traffic date picker. 

### To test

Enable the feature flag: swipe right on any screen (h/t to @staskus for telling me about this), tap Feature Flags, enable `Stats Traffic Tab`. Then open the Stats Traffic tab.

#### Day selection
1. Go back and forth between different days 
2. You shouldn't be able to go to future dates

#### Week selection
1. Switch to "By week"
2. Go back and forth between different calendar weeks (Monday to Sunday, inclusive)
3. The current week may include future days, but it should prevent you from going to the next week after that

#### Month selection
1. Switch to "By month"
2. Go back and forth between different calendar months (e.g. Jan 1-31st)
3. The current month may include future days, but it should prevent you from going to the next month after that

#### Year selection
1. Switch to "By year"
2. Go back and forth between different calendar years (e.g. Jan 1 to Dec 31, 2024)
3. The current year may include future days, but it should prevent you from going to the next year after that

#### Switching periods
1. Switch to "By week"
2. Go to a given week
3. Switch to "By day"
4. The date the picker lands on should be the date you were in beforehand



## Regression Notes
1. Potential unintended areas of impact

The changes here are mostly isolated to the Traffic tab.

5. What I did to test those areas of impact (or what existing automated tests I relied on)

Not applicable.

6. What automated tests I added (or what prevented me from doing so)

I manually tested the new date picker and added unit tests.

PR submission checklist:

- [x] I have completed the Regression Notes.
- [x] I have considered adding unit tests for my changes.
- [x] I have considered adding accessibility improvements for my changes.
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

UI Changes testing checklist:
- [x] Portrait and landscape orientations.
- [x] Light and dark modes.
- [x] Fonts: Larger, smaller and bold text.
- [x] High contrast.
- [x] VoiceOver.
- [x] Languages with large words or with letters/accents not frequently used in English.
- [x] Right-to-left languages. (Even if translation isn’t complete, formatting should still respect the right-to-left layout)
- [x] iPhone and iPad. 
- [ ] Multi-tasking: Split view and Slide over. (iPad)

Commit:
Add Stats Traffic date picker

dangermattic commented:
<!--
  1 Error: PR requires at least one label...
  0 Warnings
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">PR requires at least one label.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

guarani commented:
👋 @staskus, a heads-up: the date picker is working functionally, but visually isn't quite there yet. It also needs localization, testing, etc.

staskus commented on WordPress/Classes/ViewRelated/Stats/Traffic/TrafficView.swift:
```diff
@@ -0,0 +1,16 @@
+import SwiftUI
+
+struct TrafficView: View {
```
As I understand this one is just for testing right now?

We still need to decide if we agree to reuse existing VMs and VCs (https://github.com/wordpress-mobile/WordPress-iOS/issues/22383#issue-2078810650). For now, I also build bar chart cell not thinking too much about other setups.

guarani replied:
Yes, it's just for testing 👍

dangermattic commented:
<!--
  0 Errors
  1 Warning: PR is not assigned to a milest...
  0 Messages
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is not assigned to a milestone.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

staskus commented on WordPress/Classes/ViewRelated/Stats/Traffic/StatsTrafficDatePickerViewModel.swift:
```diff
@@ -0,0 +1,108 @@
+
+class StatsTrafficDatePickerViewModel: ObservableObject {
+
+    enum Period: String, CaseIterable {
```
We do have `StatsPeriodUnit` defined in `WordPressKit`. It could be more convenient to work with a single type, especially since we'll need to communicate changes made by this picker with other views. Wdyt?

staskus commented on WordPress/Classes/ViewRelated/Stats/Traffic/StatsTrafficDatePickerViewModel.swift:
```diff
@@ -0,0 +1,108 @@
+
+    private let formatter: DateFormatter = {
+        let formatter = DateFormatter()
+        formatter.dateFormat = "MMM d, yyyy"
```
I agree this is where we should be more flexible and have different date formats depending on the selected Period type.

staskus commented on WordPress/WordPressTest/StatsTrafficTests.swift:
```diff
@@ -0,0 +1,87 @@
+import XCTest
+@testable import WordPress
+
+final class StatsTrafficTests: XCTestCase {
```
nit-pick but I think we can have a separate test file for this picker

```suggestion
final class StatsTrafficDatePickerViewModel: XCTestCase {
```

staskus commented on WordPress/Classes/ViewRelated/Stats/Traffic/StatsTrafficDatePickerViewModel.swift:
```diff
@@ -0,0 +1,108 @@
+        return DateInterval(start: newStartDate, end: newEndDate)
+    }
+
+    private static func calculateDateInterval(for period: Period, oldDateInterval: DateInterval) -> DateInterval {
```
> Improve timezone handling

I agree that date-related logic could get complicated and require a thorough review and corner-case testing.

Have you considered using `StatsPeriodHelper` for these calculations? 

- It's already used by the current header
- It's covered by some tests (`StatsPeriodHelperTests`)
- It has a history of corner-case fixes
- - It takes into account the current date for site (`currentDateForSite` and `normalizedForSite`) (Of course we can pass `currendDateForSite` to this VM initializer)

But there could be arguments for writing a clean version of `StatsPeriodHelper` logic. 

>  If it's Saturday in the site's timezone but still Friday night for the user, should the picker allow the user to pick Saturday? I think it make sense

I agree it makes sense. Saturday data for the site is already being collected.

It would be good to add tests for cases such as this in `StatsTrafficDatePickerViewModel`.

staskus commented on WordPress/Classes/ViewRelated/Stats/Traffic/StatsTrafficDatePickerViewModel.swift:
```diff
@@ -0,0 +1,108 @@
+        return nextDateInterval.start <= now
+    }
+
+    init(now: Date = Date()) {
```
- If I understand correctly, would the idea be to pass the `currentDateForSite` to `StatsTrafficDatePickerViewModel` initializer to take into account the current time of the site?

staskus commented on WordPress/Classes/ViewRelated/Stats/Traffic/StatsTrafficDatePickerViewModel.swift:
```diff
@@ -0,0 +1,108 @@
+    }
+
+    func formattedCurrentInterval() -> String {
+        if currentDateInterval.duration < 24 * 60 * 60 {    // different format when showing a single day vs. a period such as a week
```
Agree, as I mentioned in the main review comment, it could also be different for `month` and `year`. We only need a period of formatting for a week. Others can be just the name of a day, month, or year.

staskus commented on WordPress/Classes/ViewRelated/Stats/Traffic/StatsTrafficDatePickerViewModel.swift:
```diff
@@ -0,0 +1,108 @@
+        case .day:
+            return DateInterval(start: anchorDate, end: anchorDate)
+        case .week:
+            guard let weekInterval = calendar.dateInterval(of: .weekOfYear, for: anchorDate),    // Spans 8 days since end date is exclusive
```
I was looking at this but it seems to be correct 🤔 

There's `weekIncludingDate` logic within `StatsPeriodHelper` which adds 6 days instead of adding one week and subtracting `-1`. Also the whole `StatsPeriodHelper` uses `normalizedDate` that erases seconds, and minutes, so it's enough to add 6 days to move into the last day, instead of subtracting one second.

staskus reviewed (commented):
> Shorten the date format, e.g. "Jan 14, 2024 - Jan 20, 2024" is too long

1. Agree. I think we should follow what is shown on designs IqhXWz3Iir7RMb5XH5gGfZ-fi-97_5069:

<img width="393" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/2bf1f032-026f-4b80-8c2e-c47956a0ead9">
<img width="392" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/ef547d62-f173-4c47-8945-e7aec70322c5">
<img width="400" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/377c0b70-96cb-4f4b-854c-de7a99229b4b">
<img width="399" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/48ed478d-88ae-435e-afff-c868ddc68757">

2. I also noticed that By Month, and By Year uses the date range instead of a single date, that could be fixed as well and would allow dates to fit:

<img width="396" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/9efb1cda-025e-4976-83a3-514c1cef50d7">
<img width="392" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/c1a993a2-f8f3-4ce7-8113-8d76a7371e40">

3. Nit-pick. I think the space between chevron and date text could be a bit larger (`Length.Padding.single`) according to designs

guarani commented:
👋 Thanks for the review comments, @staskus! I should have considered re-using some of the existing logic, so I'll see what can be done there. 
Yes, you're right about the date format in the designs; it shows how to represent the different periods, so I will update it to match.

Commit:
Merge branch 'trunk' into task/22382-stats-traffic-create-date-picker

mokagio commented:
I'm going to bump this to the next release because we'll be code freezing [24.3](https://github.com/wordpress-mobile/WordPress-iOS/milestone/271) today and it seems like this has not been approved yet.

It doesn't seem the case, but if it's very important that this makes it into this release, let me know and we'll organize a new beta once ready.

Commit:
Merge branch 'trunk' into task/22382-stats-traffic-create-date-picker

Commit:
Replace Period with StatsPeriodUnit

Commit:
Fix date period formatting

Commit:
Integrate date picker

Commit:
Adjust to designs

Commit:
Merge branch 'trunk' into task/22382-stats-traffic-create-date-picker

Commit:
Clean up code formatting

Commit:
Add Traffic date picker tracking

Commit:
Use site date & existing helpers

Commit:
Fixed date formatting

Commit:
Fix date navigation

Commit:
Refactor to simplify naming

Commit:
Updated tests

guarani commented:
👋 Hey @staskus, this should be working now. Thanks for all the great feedback; I think I've incorporated it all in. 

I added analytics, fixed the design and date formats, connected the date picker to the site's date, and refactored to use the existing date methods.

I haven't gone back and re-tested it all yet. If you have time, I'd appreciate your testing of this, but if not, I'll come back here and re-test everything tomorrow.

staskus commented on WordPress/Classes/ViewRelated/Stats/Period Stats/SiteStatsPeriodTableViewController.swift:
```diff
@@ -172,14 +156,10 @@ private extension SiteStatsPeriodTableViewController {
 
         refreshControl.endRefreshing()
-        tableHeaderView?.animateGhostLayers(viewModel.isFetchingChart() == true)
+//        tableHeaderView?.animateGhostLayers(viewModel.isFetchingChart() == true)
```
This can be removed then. This change means that we can navigate between the dates when data is loading which I think is fine. StatsPeriodStore cancels ongoing requests when we refresh data.

staskus commented on WordPress/Classes/ViewRelated/Stats/SiteStatsDashboardViewController.swift:
```diff
@@ -1,4 +1,5 @@
 import UIKit
+import SwiftUI
```
nitpick; unnecessary import?

guarani replied:
Good catch, I forgot to remove it after adding it for testing earlier. Removed now

staskus commented on WordPress/Classes/ViewRelated/Stats/Traffic/StatsTrafficDatePickerView.swift:
```diff
@@ -0,0 +1,91 @@
+                Image(systemName: "chevron.left")
+                    .imageScale(.small)
+                    .foregroundColor(Color.DS.Foreground.secondary)
+                    .flipsForRightToLeftLayoutDirection(true)
```
👍

staskus commented on WordPress/Classes/ViewRelated/Stats/Traffic/StatsTrafficDatePickerView.swift:
```diff
@@ -0,0 +1,91 @@
+                    .imageScale(.small)
+                    .foregroundColor(isNextDisabled ? disabledColor : enabledColor)
+                    .flipsForRightToLeftLayoutDirection(true)
+            }.disabled(isNextDisabled)
```
nitpick, modifier alignment. Here and with `.padding(..)
```suggestion
            }
            .disabled(isNextDisabled)
```

staskus commented on WordPress/Classes/ViewRelated/Stats/Traffic/StatsTrafficDatePickerView.swift:
```diff
@@ -0,0 +1,91 @@
+                .foregroundColor(Color.DS.Foreground.primary)
+                .lineLimit(1)
+
+            Spacer().frame(width: Length.Padding.single)
```
Although designs have the spacing of `.single` (8) it visually has a larger spacing due to wider padding of buttons, both between text and arrow and between arrows themselves. So either we can increase the padding of arrow buttons or increase the spacing.

<img width="118" alt="image" src="https://github.com/wordpress-mobile/WordPress-iOS/assets/4062343/3ec978ab-2b7d-4ba4-a5a3-df8455081f92">

```suggestion
            Spacer().frame(width: Length.Padding.split)
```

guarani replied:
Good idea!

staskus commented on WordPress/Classes/ViewRelated/Stats/Period Stats/SiteStatsPeriodTableViewController.swift:
```diff
@@ -193,14 +173,12 @@ private extension SiteStatsPeriodTableViewController {
         }
         addViewModelListeners()
-        viewModel?.refreshTrafficOverviewData(withDate: selectedDate, forPeriod: selectedPeriod)
+        viewModel.refreshTrafficOverviewData(withDate: datePickerViewModel.date, forPeriod: datePickerViewModel.period)
```
According to requirements (https://github.com/wordpress-mobile/jetpack-issue-repo/issues/1), we need to:

> Remember the last selected tab and date range (as we do currently with Day/Week/Month/Year tabs)

It's not working with this PR but I have implemented it in https://github.com/wordpress-mobile/WordPress-iOS/pull/22695 and the same solution should continue working.

guarani replied:
Thanks for catching and fixing that!

staskus reviewed (commented):
Thanks, looks and works well! Only left a couple of comments. 

## Day selection
- [x] Go back and forth between different days
- [x] You shouldn't be able to go to future dates

## Week selection

Switch to "By week"
- [x] Go back and forth between different calendar weeks (Monday to Sunday, inclusive)
- [x] The current week may include future days, but it should prevent you from going to the next week after that

## Month selection
- [x] Switch to "By month"
- [x] Go back and forth between different calendar months (e.g. Jan 1-31st)
- [x] The current month may include future days, but it should prevent you from going to the next month after that

## Year selection
- [x] Switch to "By year"
- [x] Go back and forth between different calendar years (e.g. Jan 1 to Dec 31, 2024)
The current year may include future days, but it should prevent you from going to the next year after that

## Switching periods
- [x] Switch to "By week"
- [x] Go to a given week
- [x] Switch to "By day"
- [x] The date the picker lands on should be the date you were in beforehand

## Other tests

- [x] Analytics
- [x] RTL Language
- [x] Dark mode
- [x] Larger fonts
- [x] Different orientations
- [x] Voiceover

staskus commented on WordPress/Classes/ViewRelated/Stats/Traffic/StatsTrafficDatePickerView.swift:
```diff
@@ -0,0 +1,91 @@
+            let enabledColor = Color.DS.Foreground.secondary
+            let disabledColor = enabledColor.opacity(0.5)
+
+            Button(action: {
```
Feedback on UX: Since switching between dates is a crucial part of UX, I think that the arrows are too small and coupled with the small spaces between them make it inconvenient to tap on a button.

guarani replied:
Good point. I could simply increase the chevron size here. Alternatively, instead of the SF Symbol, I could use the `chevron-right` image from Figma, which includes padding.
I'm going to follow up in a separate PR. I think there might be a couple of other adjustments to do (e.g. when showing months from previous years, adding the year would be super helpful).

staskus replied:
Sounds good, lets leave it for improvement PR. 👍

Commit:
Fix code alignment

Co-authored-by: Povilas Staskus <povilas.staskus@automattic.com>

Commit:
Adjusted date picker navigation button spacing

Co-authored-by: Povilas Staskus <povilas.staskus@automattic.com>

Commit:
Clean up code

staskus commented:
I found a corner-case issue that combines these two requirements:
> The current month may include future days, but it should prevent you from going to the next month after that
> The date the picker lands on should be the date you were in beforehand

1. Select By month
2. Come back a couple of months to November, 2023
3. Select By Year
4. Confirm 2023 is shown
5. Go forward to 2024
6. Select by month
7. Now the data loads for November, 2024 which is an unexistent future Month

guarani commented:
Thanks for catching that. I've made a separate issue for it: https://github.com/wordpress-mobile/jetpack-issue-repo/issues/19

Enabling auto-merge now
